src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1499 @@ void Init()
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_TypeTest()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { IsOk: bool }:
+                return Value; // 1
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (16,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(16, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_StateAfterSwitch()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { IsOk: true }:
+                break;
+            default:
+                throw null!;
+        }
+        return Value;
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_WithBoolPattern()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public bool IsIrrelevant { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { IsOk: true, IsIrrelevant: true }:
+                return Value;
+            case { IsOk: true, IsIrrelevant: not true }:
+                return Value;
+            default:
+                return Value; // 1
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (21,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(21, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_WithNullPattern()
```

Comment:
>MemberNotNullWhenTrue_RecursivePattern_WithNullPattern [](start = 20, length = 54)

Does this test cover a distinct scenario from the previous test?

Reply:
The above test follows the `IsOk` check with a boolean test, whereas this one follows it with a null test. I was trying to cover difference code paths, but I don't think it ended mattering. I could remove.

---
In reply to: [573105354](https://github.com/dotnet/roslyn/pull/51001#discussion_r573105354) [](ancestors = 573105354)